File: kotlin/v2g-keystore/src/main/kotlin/cloud/charging/v2g/keystore/V2GCsr.kt
package cloud.charging.v2g.keystore

import java.io.OutputStream
import org.bouncycastle.asn1.ASN1Integer
import org.bouncycastle.asn1.DERSequence
import org.bouncycastle.asn1.ASN1EncodableVector
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
import org.bouncycastle.operator.ContentSigner
import org.bouncycastle.operator.DefaultSignatureAlgorithmIdentifierFinder
import org.bouncycastle.pkcs.PKCS10CertificationRequest
import org.bouncycastle.pkcs.PKCS10CertificationRequestBuilder
import java.math.BigInteger

/**
 * Builds a PKCS#10 certificate signing request for a key this app holds.
 *
 * ## Why this exists in this module rather than next to the certificates
 *
 * A CSR proves possession of a private key, so it must be **signed by that key** — and if the key
 * lives in a secure element, its bytes never leave. That makes CSR generation the first real
 * customer of [V2GSigner]: the request is assembled, the octets are handed out, a signature comes
 * back. Had `PncEvccOptions` and this both taken a raw key, hardware backing would have been
 * impossible in two places instead of one.
 *
 * ## The conversion that is easy to get wrong
 *
 * [V2GSigner] returns **raw `r‖s`**, because that is what ISO 15118 puts on the wire and the field
 * is sized for it. PKCS#10 wants the **DER** form — `SEQUENCE { INTEGER r, INTEGER s }` — like every
 * other X.509 structure. So the signature has to be converted here, and a CSR carrying a raw
 * signature is not malformed in any way a parser would notice quickly: it simply fails to verify at
 * whichever CA receives it, which is a miserable place to find out.
 */
object V2GCsr {

    /**
     * @param subject an RFC 4514 / RFC 2253 distinguished name, e.g. `CN=DE8AA1A2B3C4D5`
     * @return the DER-encoded CertificationRequest
     */
    fun build(subject: String, signer: V2GSigner): ByteArray {

        require(signer.curve == V2GKeyCurve.P256) {
            "only P-256 CSRs are supported so far; ${signer.curve} needs its own signature algorithm " +
            "identifier and a matching r‖s width."
        }

        val builder = PKCS10CertificationRequestBuilder(
            X500Name(subject),
            SubjectPublicKeyInfo.getInstance(signer.publicKeyDer))

        return builder.build(RawEcdsaContentSigner(signer)).encoded
    }

    /** Reads a CSR back, for tests and for showing a user what is about to be sent. */
    fun parse(der: ByteArray): PKCS10CertificationRequest = PKCS10CertificationRequest(der)

    /**
     * A [ContentSigner] over a [V2GSigner]: it accumulates the bytes BouncyCastle wants signed, hands
     * them to the signer, and converts the raw `r‖s` that comes back into the DER form PKCS#10 wants.
     */
    private class RawEcdsaContentSigner(private val signer: V2GSigner) : ContentSigner {

        private val buffer = java.io.ByteArrayOutputStream()

        override fun getAlgorithmIdentifier(): AlgorithmIdentifier =
            DefaultSignatureAlgorithmIdentifierFinder().find("SHA256WITHECDSA")

        override fun getOutputStream(): OutputStream = buffer

        override fun getSignature(): ByteArray = derFromRawRs(signer.signature(buffer.toByteArray()))
    }

    /**
     * `r‖s` → `SEQUENCE { INTEGER r, INTEGER s }`.
     *
     * The two halves are fixed-width unsigned big-endian; DER integers are signed and minimally
     * encoded, so a leading zero appears exactly when the high bit is set and disappears otherwise.
     * `BigInteger(1, bytes)` gets both right, which is why it is used rather than any byte fiddling.
     */
    internal fun derFromRawRs(raw: ByteArray): ByteArray {
        require(raw.size % 2 == 0) { "a raw r‖s signature has two equal halves; got ${raw.size} bytes" }
        val half = raw.size / 2
        val r = BigInteger(1, raw.copyOfRange(0, half))
        val s = BigInteger(1, raw.copyOfRange(half, raw.size))
        return DERSequence(ASN1EncodableVector().apply {
            add(ASN1Integer(r))
            add(ASN1Integer(s))
        }).encoded
    }
}

File: swift/Sources/V2GKeystore/V2GCsr.swift
import CryptoKit
import Foundation
import SwiftASN1
import X509

/// Builds a PKCS#10 certificate signing request for a key this app holds.
///
/// ## Why a CSR is the first real customer of ``V2GSigner``
///
/// A CSR proves possession of a private key, so it must be **signed by that key** — and if the key
/// lives in a secure element its bytes never leave. The request is assembled, the octets are handed
/// out, a signature comes back. Had this taken a private key, hardware-backed credentials would have
/// been impossible in a second place as well as the first.
///
/// ## Two things worth knowing about the implementation
///
/// **`swift-certificates` cannot quite do this on its own.** It offers an initialiser taking a
/// pre-computed signature, which is exactly the external-signer shape — but the bytes that signature
/// must cover (`CertificationRequestInfo`) are `@usableFromInline internal`, so there is no public
/// way to ask for them. The way through is to build the request once with a placeholder signature,
/// serialise it, and take the first element of the outer `SEQUENCE`, which by PKCS#10's own
/// definition *is* those bytes. Then sign and build again. BouncyCastle, which the Kotlin half uses,
/// takes an external `ContentSigner` directly and needs none of this.
///
/// **The signature form changes here.** ``V2GSigner`` returns raw `r‖s`, because that is what ISO
/// 15118 puts on the wire and the field is sized for it. X.509 and PKCS#10 want the DER form,
/// `SEQUENCE { INTEGER r, INTEGER s }`. A CSR carrying a raw signature is not malformed in a way any
/// parser notices — it simply fails to verify at whichever CA receives it.
public enum V2GCsr {

    /// - Parameters:
    ///   - subject: the distinguished name to request, e.g. `CN=DE8AA1A2B3C4D5`
    ///   - signer: the key proving possession; only P-256 so far
    /// - Returns: the DER-encoded CertificationRequest
    public static func build(subject: DistinguishedName, signer: any V2GSigner) throws -> [UInt8] {

        guard signer.curve == .p256 else {
            throw V2GKeyError.unsupportedCurve(signer.curve)
        }

        let publicKey = try Certificate.PublicKey(
            P256.Signing.PublicKey(derRepresentation: Data(signer.publicKeyDer)))

        // Round one: a placeholder signature, only so the request can be serialised and the bytes
        // that a real signature must cover read back out of it.
        let placeholder = try Certificate.Signature(
            signatureAlgorithm: .ecdsaWithSHA256,
            signatureBytes: ASN1BitString(bytes: ArraySlice(placeholderDerSignature)))

        let draft = try CertificateSigningRequest(
            version: .v1, subject: subject, publicKey: publicKey,
            attributes: CertificateSigningRequest.Attributes(),
            signatureAlgorithm: .ecdsaWithSHA256, signature: placeholder)

        var serializer = DER.Serializer()
        try serializer.serialize(draft)
        let info = try firstElement(ofSequence: serializer.serializedBytes)

        // Round two: the real signature, over exactly those bytes.
        let raw = try signer.signature(over: info)
        let der = try P256.Signing.ECDSASignature(rawRepresentation: Data(raw)).derRepresentation

        let signature = try Certificate.Signature(
            signatureAlgorithm: .ecdsaWithSHA256,
            signatureBytes: ASN1BitString(bytes: ArraySlice(Array(der))))

        let request = try CertificateSigningRequest(
            version: .v1, subject: subject, publicKey: publicKey,
            attributes: CertificateSigningRequest.Attributes(),
            signatureAlgorithm: .ecdsaWithSHA256, signature: signature)

        var out = DER.Serializer()
        try out.serialize(request)
        return out.serializedBytes
    }

    /// Any well-formed 64-byte ECDSA signature in DER, used only to make the draft serialisable. Its
    /// value never reaches a caller: the draft exists to be measured, not to be sent.
    private static var placeholderDerSignature: [UInt8] {
        Array(try! P256.Signing.ECDSASignature(rawRepresentation: Data(repeating: 1, count: 64))
                  .derRepresentation)
    }

    /// The first TLV inside a definite-length DER `SEQUENCE`, whole — header and content.
    ///
    /// Deliberately a small explicit walk rather than a parse-and-reserialise: what is wanted is the
    /// *original* bytes, and a value that round-trips through any encoder is not guaranteed to come
    /// back identical. A signature over re-encoded bytes would verify here and nowhere else.
    internal static func firstElement(ofSequence der: [UInt8]) throws -> [UInt8] {

        var index = 0

        func readHeader() throws -> (contentStart: Int, length: Int) {
            guard index + 1 < der.count else { throw V2GCsrError.truncated }
            index += 1                                   // tag
            let first = Int(der[index]); index += 1      // length

            if first < 0x80 { return (index, first) }

            let count = first & 0x7F
            guard count > 0, count <= 4, index + count <= der.count else { throw V2GCsrError.truncated }

            var length = 0
            for _ in 0 ..< count { length = (length << 8) | Int(der[index]); index += 1 }
            return (index, length)
        }

        _ = try readHeader()                             // the outer SEQUENCE
        let elementStart = index
        let (contentStart, length) = try readHeader()    // the first element

        guard contentStart + length <= der.count else { throw V2GCsrError.truncated }
        return Array(der[elementStart ..< contentStart + length])
    }
}

public enum V2GCsrError: Error, CustomStringConvertible {
    case truncated

    public var description: String {
        switch self {
        case .truncated: return "the DER ended in the middle of a value."
        }
    }
}
